fix(cli): #1585 retry file copy on transient Windows file locks (EBUSY/EPERM)

SiddarthaKarri opened the pull request:
## Related Issue

Resolves #1585 

## Summary of Changes

This PR reduces intermittent Windows CI failures caused by transient file-lock/resource-locked errors during file copy operations (EBUSY / EPERM / EACCES) by adding retry logic to the primary copy code paths.

What I changed
- Added an async retrying copy helper:
  - fs-utils.js — `copyFileWithRetry(source, target, { attempts, baseDelay })`
- Wired the helper into runtime copy paths:
  - copy.js — use `copyFileWithRetry` for file copies in the recursive copy lifecycle; improved copy failure logging.
  - plugin-standard-css.js — use `copyFileWithRetry` when copying resolved assets into the output (call is non-blocking from the synchronous bundler; errors are logged).
- Made the CLI init/template copier more robust:
  - util.js — added `copyFileSyncWithRetry` to retry synchronous template copy operations (keeps `copyTemplate` sync).

Checklist
- [x] Add retry helper for async copy
- [x] Replace copy calls in the core copy lifecycle
- [x] Replace copy call in CSS asset handling
- [x] Add sync retry for init template copying
- [ ] (optional) Convert CSS bundler to async and await copy operations (see notes)

Files changed
- Added: fs-utils.js — async copy + retry
- Updated: copy.js — use retry helper; better error logging
- Updated: plugin-standard-css.js — use retry helper for bundled asset copying
- Updated: util.js — added `copyFileSyncWithRetry` and used it

SiddarthaKarri committed:
fix(cli): retry file copy on transient Windows file locks (EBUSY/EPERM)

KaiPrince commented on packages/init/src/util.js:
```diff
@@ -3,6 +3,31 @@ import os from "node:os";
+        const delay = baseDelay * Math.pow(2, i);
+        const waitUntil = Date.now() + delay + Math.floor(Math.random() * baseDelay);
+        while (Date.now() < waitUntil) {
+          // busy wait — this is synchronous intentionally
```
This will consume a lot of unnecessary CPU and drain the battery if run on a user’s device. I understand it’s being run in a synchronous context, but I’m looking for ways we can avoid this. There’s a thread-sleep module that seems to have found a clever trick with node’s execSync function
https://github.com/ForbesLindesay/thread-sleep/blob/master/index.js

another thing that *maybe* could help is this fancy Atomics thing I found. Although I admit I haven’t finished looking into it so maybe not. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/pause

KaiPrince commented on packages/cli/src/lib/fs-utils.js:
```diff
@@ -0,0 +1,33 @@
+      // Only retry for transient EBUSY / EPERM on Windows or generic resource busy errors
+      if ((err.code === 'EBUSY' || err.code === 'EPERM' || err.code === 'EACCES') && i < attempts - 1) {
+        const delay = baseDelay * Math.pow(2, i);
+        // add small jitter
```
I’m curious, why would jitter be useful here? It seems cool, but when I actually think about it I can’t think of an actual use case for it

thescientist13 commented:
Thanks @SiddarthaKarri , appreciate the PR and apologies in the delay for reviewing this.

Aside from some of the feedback provided by @KaiPrince , one thing I was mostly curious from the original issue was the "why", so to speak.  Is the issue that Windows can't handle so many file copies at once?  For the project being built here in our CI, I don't _think_ it's that many files?
https://github.com/ProjectEvergreen/greenwood/actions/runs/18298962229/job/52102995004
<details>
  <pre>
    copying file... .greenwood/manifest.json
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/webcomponents-loader.js
    copying directory... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/
    copying file... /home/runner/work/greenwood/greenwood/node_modules/lit/polyfill-support.js
    copying directory... src/assets/
    copying file... src/favicon.ico
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-ce.js
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-ce.js.map
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-pf_dom.js
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-pf_dom.js.map
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-pf_js.js
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-pf_js.js.map
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-sd-ce-pf.js
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-sd-ce-pf.js.map
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-sd-ce.js
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-sd-ce.js.map
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-sd.js
    copying file... /home/runner/work/greenwood/greenwood/node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-sd.js.map
    copying file... src/assets/evergreen.svg
    copying file... src/assets/getting-started-netlify-config.png
    copying file... src/assets/getting-started-repo-styled.png
    copying file... src/assets/getting-started-repo-unstyled-partial.png
    copying file... src/assets/gh-pages-branch-commits.png
    copying file... src/assets/gh-pages-branch.png
    copying file... src/assets/graphql-playground.png
    copying file... src/assets/greenwood-getting-started-repo-optimized.webp
    copying file... src/assets/greenwood-logo-1000w.webp
    copying file... src/assets/greenwood-logo-1500w.webp
    copying file... src/assets/greenwood-logo-300w.webp
    copying file... src/assets/greenwood-logo-500w.webp
    copying file... src/assets/greenwood-logo-750w.webp
    copying file... src/assets/greenwood-logo-og.png
    copying file... src/assets/greenwood-starter-presentation.png
    copying file... src/assets/link.png
    copying file... src/assets/netlify-admin.png
    copying file... src/assets/netlify-cms.jpg
    copying file... src/assets/netlify-create-new.png
    copying file... src/assets/netlify-deploy.png
    copying file... src/assets/netlify-git-gateway.png
    copying file... src/assets/netlify-invite.png
    copying file... src/assets/netlify-registration.png
    copying file... src/assets/netlify-workflow.png
    copying file... src/assets/nodejs.png
    copying file... src/assets/repo-github-pages-config.png
    copying file... src/assets/serverless.webp
    copying file... src/assets/simple.png
    copying file... src/assets/web-components-browser-support.png
    copying file... src/assets/webcomponents.jpg
    copying file... src/assets/fonts/source-sans-pro-v13-latin-regular.ttf
    copying file... src/assets/fonts/source-sans-pro-v13-latin-regular.woff
    copying file... src/assets/fonts/source-sans-pro-v13-latin-regular.woff2
    copying file... src/assets/fonts/source-sans-pro.css
    copying file... src/assets/blog-images/dev-cache-step1.png
    copying file... src/assets/blog-images/dev-cache-step2.png
    copying file... src/assets/blog-images/dev-cache-step3.png
    copying file... src/assets/blog-images/dev-cache-step4.png
    copying file... src/assets/blog-images/full-stack-web-components.webp
    copying file... src/assets/blog-images/hud.png
    copying file... src/assets/blog-images/init-scaffolding.png
    copying file... src/assets/blog-images/wcc-logo.png
    copying file... src/assets/blog-images/not-found.png
    copying file... src/assets/blog-images/ssr.webp
  </pre>
</details>

So what limit / threshold are we hitting, and if it's a Node issue or Windows issue?  (docs / links of others having a similar issue would be great here!)

I ask because it will help give me context on if the solution here is an appropriate one for the given problem.  I understand the PR is implementing some retry logic, but is there a way to avoid the original issue altogether; maybe copying less files at once or something else?  Maybe just making copying synchronous (again)?  🤔 

Thanks!

SiddarthaKarri committed:
Merge branch 'master' into chore/issue-1585-add-copyfile-retry

SiddarthaKarri committed:
fix(cli): retry file copy on transient Windows file locks (EBUSY/EPERM)

SiddarthaKarri commented:
Thanks for the feedback — I pushed an update. I removed the busy-wait, restored init’s sync copy, and added bounded-concurrency to the copy lifecycle (files=8, locations=4, plugins=2) while keeping an async retry wrapper for transient EBUSY/EPERM errors. Could someone re-run the Windows CI for this branch? If flakes persist I’ll add verbose retry logging or convert the CSS bundler to async next.

SiddarthaKarri committed:
test: add safeTeardown helper and retry teardown in failing init test to reduce intermittent EBUSY on Windows CI

SiddarthaKarri committed:
chore(init+cli): add copy retry and safe teardown to reduce Windows EBUSY/EPERM flakes (issue #1585)

thescientist13 commented on packages/init/test/cases/develop.default/develop.default.spec.js:
```diff
@@ -93,8 +93,11 @@ describe("Initialize a new Greenwood project: ", function () {
+  after(async function () {
     runner.stopCommand();
-    runner.teardown([initOutputPath]);
+    // give the process a moment to release file handles (helps on Windows)
```
I don't think we want this, this should be taken care of by #1573 which includes a fix to the test runner around `teardown` exit conditions

thescientist13 reviewed (commented):
Unless I have missed it somewhere in the comments, I still don't see any insight into why the issue may be happening in the first place, which I would still like to get a better understanding of first before going back and forth on solutions here, since without a firm understanding of why, I think any solution would just be in "spirit" only.

To add some food for thought, one thing I would like to know is why are we getting this on a copy operation?  Is it because we copy files `async` and thus they all copy "out of order" and so when each invocation tries to make it's output directory, that's what causes the issue?

Would making it sync (like it was before) be a solution?  That said, I think this issue pre-dates that refactor though.

So yeah, until we can establish a better understanding of the root cause of this issue, I'm going to put this PR into draft until then.

Thanks again for getting things started, let us know if you have any questions and always happy to help guide the conversation.  (feel free to join the Discord!)

thescientist13 reviewed (commented):
I took a look into this and think I may have [found the root cause]
(https://github.com/ProjectEvergreen/greenwood/pull/1728). (please feel free to review / comment there 🙏 )

There was an async process that should have been sync.  Same file, different solution, so thinking I might go with that one.  Will play around with it a bit and see if it holds up. 🤞

thescientist13 commented:
Closing in favor of #1728 / #1750
